.pre-commit/pre-commit.Dockerfile

```diff
@@ -1,11 +1,11 @@
-FROM nikolaik/python-nodejs:python3.9-nodejs16@sha256:e858a798bf7ec2f4174e7ff756e6c83eb123b687e451a9bf8aced59fa9d2be75 as base
+FROM nikolaik/python-nodejs:python3.12-nodejs21 as base
```

Comment:
Could we also pin the exact image hash too?

Reply:
Discussed offline -- digests are tied to specific architectures so will just leave the tag for now.